terraform/main.tf

```diff
@@ -393,15 +396,144 @@ resource "aws_security_group" "c11-railway-tracker-dashboard-sg-tf" {
+        security_groups  = [aws_security_group.c11-railway-tracker-dashboard-sg-tf.id] 
         assign_public_ip = true
     }
 }
+
+# --------------- INCIDENT: LAMBDA & EVENT BRIDGE
+
+# IAM Role for Lambda execution
+resource "aws_iam_role" "lambda_execution_role" {
+  name = "lambda_execution_role"
+
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action    = "sts:AssumeRole"
+        Effect    = "Allow"
+        Principal = {
+          Service = "lambda.amazonaws.com"
+        }
+      }
+    ]
+  })
+}
+
+# IAM Policy for Lambda execution role
+resource "aws_iam_role_policy" "lambda_execution_policy" {
+  name = "lambda_execution_policy"
+  role = aws_iam_role.lambda_execution_role.id
+
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action = [
+          "logs:CreateLogGroup",
+          "logs:CreateLogStream",
+          "logs:PutLogEvents"
+        ]
+        Effect   = "Allow"
+        Resource = "*"
+      },
+      {
+        Action   = "dynamodb:Query"
+        Effect   = "Allow"
+        Resource = "*"
+      }
+    ]
+  })
+}
+
+# Lambda Function Resource
+resource "aws_lambda_function" "c11_trainwreck_national_rail" {
+  function_name = "c11-trainwreck-national-rail"
+  role          = aws_iam_role.lambda_execution_role.arn
+  image_uri     = "129033205317.dkr.ecr.eu-west-2.amazonaws.com/c11-trainwreck-national:latest"
+  package_type  = "Image"
+  timeout       = 180
+
+  environment {
+    variables = {
+      ACCESS_KEY_ID         = var.AWS_ACCESS_KEY,
+      SECRET_ACCESS_KEY     = var.AWS_SECRET_KEY,
+      DB_IP                 = var.DB_IP,
+      DB_NAME               = var.DB_NAME,
+      DB_USERNAME           = var.DB_USERNAME,
+      DB_PASSWORD           = var.DB_PASSWORD,
+      DB_PORT               = var.DB_PORT,
+      NATIONAL_RAIL_API_KEY = var.NATIONAL_RAIL_API_KEY
+    }
+  }
+
+  logging_config {
+    log_format = "Text"
+    log_group  = "/aws/lambda/c11-trainwreck-national-rail"
+  }
+
+  tracing_config {
+    mode = "PassThrough"
+  }
+}
+
+# IAM Role for AWS Scheduler
+resource "aws_iam_role" "scheduler_execution_role" {
+  name = "scheduler_execution_role"
+
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action    = "sts:AssumeRole"
+        Effect    = "Allow"
+        Principal = {
+          Service = "scheduler.amazonaws.com"
+        }
+      }
+    ]
+  })
+}
+
+# IAM Policy for AWS Scheduler role
+resource "aws_iam_role_policy" "scheduler_execution_policy" {
+  name = "scheduler_execution_policy"
+  role = aws_iam_role.scheduler_execution_role.id
+
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action   = "lambda:InvokeFunction"
+        Effect   = "Allow"
+        Resource = aws_lambda_function.c11_trainwreck_national_rail.arn
```

Comment:
Good call, but does this have to be repeated?

Reply:
no so I've split it up but I don't want to ruin the code by reusing other parts